### ISSUE TITLE
IllegalStateException: Not inside a request scope

### ISSUE THREAD

**Issue by pdurbin** [payara/Payara#532]
I attempted to switch our application (Dataverse) from Glassfish 4.1 to Payara 4.1.1.154 (in commit [1c0c7be](https://github.com/IQSS/dataverse/commit/1c0c7be6ebc332a1fd62c7d09cb09badb18d0c41)) and while our setup scripts ran fine (which was not the case when [attempted to upgraded from Glassfish 4.1 to 4.1.1](https://github.com/IQSS/dataverse/issues/2628#issuecomment-158197478)) the home page throws a 500 error and I am not able to log in:

![500_internal_server_error_-_root_dataverse_-_2015-11-19_09 26 22](https://cloud.githubusercontent.com/assets/21006/11284441/c204f956-8ed7-11e5-82d5-a586d52aaeb0.png)

When I visit the homepage I see exceptions like these in server.log...

`A system exception occurred during an invocation on EJB PermissionServiceBean, method: public edu.harvard.iq.dataverse.PermissionServiceBean$RequestPermissionQuery edu.harvard.iq.dataverse.PermissionServiceBean.on(edu.harvard.iq.dataverse.DvObject)]] ... javax.ejb.EJBException... at edu.harvard.iq.dataverse.DataversePage.init(DataversePage.java:294)... Caused by: org.jboss.weld.exceptions.WeldException: WELD-000049: Unable to invoke protected void edu.harvard.iq.dataverse.DataverseRequestServiceBean.setup() on edu.harvard.iq.dataverse.DataverseRequestServiceBean@53084b6a ... at edu.harvard.iq.dataverse.PermissionServiceBean.on(PermissionServiceBean.java:356) ... Caused by: java.lang.reflect.InvocationTargetException at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.lang.reflect.Method.invoke(Method.java:497) at org.jboss.weld.injection.producer.DefaultLifecycleCallbackInvoker.invokeMethods(DefaultLifecycleCallbackInvoker.java:98) ... 99 more Caused by: java.lang.IllegalStateException: Not inside a request scope. at jersey.repackaged.com.google.common.base.Preconditions.checkState(Preconditions.java:173)`

... and here is the full server.log (which also includes a couple attempts to log in): [payara-4.1.1.154-1c0c7be-server.log.txt](https://github.com/payara/Payara/files/39469/payara-4.1.1.154-1c0c7be-server.log.txt)

I'm struggling with what to write as the title of this issue so @smillidge (et al.) are welcome to change it. At https://github.com/payara/Payara/issues/529#issuecomment-158128834 he asked me to file an issue here if I'm seeing a difference between Glassfish 4.1.1 and Payara 4.1.1.154 and I certainly am.

To sum up:
- Glassfish 4.1: works
- Glassfish 4.1.1: doesn't work. Bean Validation problems when our setup scripts run: https://github.com/IQSS/dataverse/issues/2628#issuecomment-158197478
- Payara 4.1.1.154: doesn't work. We are able to run our setup scripts but then we can't load the homepage or log in (and who knows what else). Details and server.log above.

**Comment by abhirockzz** [comment]
I was able to reproduce this using a [simple test java ee 7 application](https://drive.google.com/file/d/0BwKHHk1eqZacMU94Sldma3U0TmM/view?usp=sharing) which does something to similar to what Dataverse application does (using @Context for injection in a non JAX-RS component such as a JSF managed bean). WAR available [here](https://drive.google.com/file/d/0BwKHHk1eqZacMU94Sldma3U0TmM/view?usp=sharing)

[DataverseRequestServiceBean](https://github.com/IQSS/dataverse/blob/28758208bef3fb713c903450067a47c0814655b1/src/main/java/edu/harvard/iq/dataverse/DataverseRequestServiceBean.java) injects an HttpServletRequest object (using @Context) without it actually being mapped as a JAX-RS root class. The presence of [ApiConfiguration](https://github.com/IQSS/dataverse/blob/28758208bef3fb713c903450067a47c0814655b1/src/main/java/edu/harvard/iq/dataverse/api/ApiConfiguration.java) results in the *java.lang.IllegalStateException: Not inside a request scope* exception - else, we would have seen a *NullPointer*

**Comment by pdurbin** [comment]
Pinging @michbarsinai since he wrote DataverseRequestServiceBean.

@abhirockzz the issue does have something to do with Bean Validation, right? From the links you posted I can't tell if you're using Bean Validation.

**Comment by abhirockzz** [comment]
No I did not mention Bean Validation. IMO, it might have something to do with the usage of @Context in a non JAX-RS component (this is what I had reproduced in the test app I posted). Sorry for not making it clear as to how to test it out. After deploying the WAR, please use this URL : 
http://<host>:<port>/payara532/faces/payara532.xhtml ..

**Comment by michbarsinai** [comment]
Sounds like we need to do another refactoring here, where the client code passes the HTTPServletRequest to the bean. This way we can have the JAX-RS and the JSF each obtaining the HTTP Request after their own manner. It's a large change, though - it's called in 55 places, all over the code. 

Another option is to look into uniform way of getting the request in the bean.

**Comment by davidweaver** [comment]
@abhirockzz thanks for the test case.
We are able to reproduce this in both Payara Server 4.1.1.154 and the latest 4.1.1.161-snapshot.

We will check the JAX-RS and JSF specs and confirm that Payara Server complies.

**Comment by mikecroft** [comment]
Internal defect ID: `PAYARA-834`

**Comment by arjantijms** [comment]
>using @context for injection in a non JAX-RS component such as a JSF managed bean

Not really sure how that could have worked, as the annotation has to be processed by the DI system, which in the case of the `DataverseRequestServiceBean` is a named request scoped bean, so CDI would process it and take care of the injection, but it obviously doesn't known about the `@Context` annotation.

A CDI extension may have been able to do some magic, so if there's indeed a CDI extension used for that, that may be a place to look at.

Additionally, CDI provides a build-in bean for `HttpServletRequest`, so you can `@Inject` the `HttpServletRequest` as well. Was there ever a specific reason to try to use `@Context` here for that?

**Comment by pdurbin** [comment]
@arjantijms thanks for taking a look and congrats on the new job with Payara! We use OmniFaces in Datavese and love it and I've definitely clicked around on https://zeef.com before. I'm afraid that I'm not well versed enough in Java EE to answer your questions but I just asked other Dataverse developers to check out what you just wrote here: https://groups.google.com/d/msg/dataverse-dev/07dyNe5BvJA/0zM-qj0LCQAJ . Everyone who installs Dataverse (25 public installations so far) is stuck on Glassfish 4.1 and we'd love for them to be able to upgrade some day.

**Comment by donsizemore** [comment]
Just a note that I seemingly successfully deployed the Dataverse 4.7 release warfile in Payara 4.1.2.172, on Ubuntu 16 (openjdk 1.8.0_131). Full server log attached in case it's helpful.

**Comment by pdurbin** [comment]
@donsizemore sure enough, I'm able to log into the server you mentioned at http://irclog.iq.harvard.edu/dataverse/2017-07-14#i_54570 (the password has since changed). Screenshot below:

![works](https://user-images.githubusercontent.com/21006/28232148-2d58e742-68bd-11e7-81b6-a2c2795369cc.png)

It's been a year since we've been discussing this issue here so I'm not sure what changed but being able to log in to the latest version of Dataverse running the latest version of Payara is a positive sign!

**Comment by pdurbin** [comment]
@smillidge sorry for not closing this out months ago. It was nice meeting you at JavaOne! Thanks again for the Payara t-shirt. 😄

**Comment by smillidge** [comment]
no problem I was just doing some housekeeping
